java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll

```diff
@@ -0,0 +1,333 @@
+private import java
+private import semmle.code.java.Collections
+private import semmle.code.java.dataflow.internal.ContainerFlow
+private import CaptureModelsSpecific as Specific
+private import CaptureModels
+
+/**
+ * A type representing instantiations of class types
+ * that has a method which returns an iterator.
+ */
+private class IterableType extends Class {
```

Comment:
There are some implicit assumptions throughout this file that an `IterableType` has exactly one type variable. This is probably fine as a starting point, as that's likely to generally be the case, but might deserve a comment.

Reply:
An IterableType can be parameterized over multiple type variables. As an example
```java
public class Stream2<T1, T2> {

    public Iterator<T1> iterator() {
        return null;
    }

    public T2 getField() {
        return null;
    }

    public T1 getElement() {
        return null;
    }
}
```
produce the models
```
p;Stream2;true;getElement;();;Argument[-1].Element;ReturnValue;value;generated
p;Stream2;true;getField;();;Argument[-1].SyntheticField[ArgType1];ReturnValue;value;generated p;Stream2;true;iterator;();;Argument[-1].Element;ReturnValue.Element;value;generated
```

I am not sure where you discovered the implicit assumptions.
If you have seen something, then it might be worth narrowing the InterableType definition, such we only accept types that are parameterized over a single type variable and where there exists an iterator method over that exact type.

Reply:
Consider this (probably an obscure corner-case, but still):
```
public class Stream2<T1, T2> {
    public Iterator<T1> iterator1() {
        return null;
    }

    public Iterator<T2> iterator2() {
        return null;
    }

    public T1 getElement1() {
        return null;
    }

    public T2 getElement2() {
        return null;
    }
}
```

Reply:
You are right. In that case, we probably shouldn't think of `Stream2` as being an interable type.
It is an easy fix.